book/src/designs/0002-parallel-verification.md

```diff
@@ -0,0 +1,426 @@
+    the `CheckpointVerifier` design.)
+* **Constraint Verification:**
+  * `process_checkpoint_range`: makes sure that the blocks in the checkpoint
+    range have an unbroken chain of previous block hashes.
+* **Contextual Verification:**
+  * *Not Required:* Verifying a chain of blocks against its checkpoints
+    confirms that the network considers those blocks valid. (Strictly, that the
+    network considered those blocks valid, up to and including the time when
+    those checkpoints were created.)
+* **In-Memory Chain Updates:**
+  * The checkpoint verifier uses an internal queue of blocks to store the
+    simple height and hash context it requires for verification.
+  * *As Above*: Although the checkpoint verifier does not require any external
+    context, Zebra needs to maintain enough context to verify the first
+    non-checkpoint block.
+  * Since there is only ever a single checkpoint chain, Zebra does not need to
+    keep any previous contexts, until it is processing the last 100
+    checkpoint blocks.
+  * Note: If any context fields are only used to verify blocks within the
+    checkpoint range, then Zebra does not need to keep that context. (For
+    example, sprout-only context.)
+* **Chain Tip Updates:**
+  * *Not Required:* Since there is only a single chain, the main chain tip is
+    the unique tip. As each checkpoint is verified, it implicitly becomes the
+    main tip.
+* **Main Chain Disk Updates:**
+  * *As Above*: Any large context that is required to verify the first
+    non-checkpoint block needs to be stored to disk.
+* **Chain Pruning:**
+  * *Not Required:* Since Zebra does not keep previous chain contexts until
+    the last 100 checkpoint blocks, it will never need to prune any old
+    contexts. The earliest checkpoint context will be pruned after the first
+    non-checkpoint block is verified. (Since there is only one checkpoint
+    chain, there are no side-chains to prune.)
+
+## Chain Context
+[chain-context]: #chain-context
+
+The `ChainContext` for a block contains the contextual information needed to
+verify the next block in the chain.
+
+There are a few exceptions to this general principle:
+* If the context is a field in recent block headers, it may be retrieved via
+  the list of recent blocks in the chain context.
+* Large state, such as unspent transaction outputs (UTXOs), is stored on disk.
+  The chain context stores deltas from recent blocks, which can be used to
+  avoid some disk state queries.
+
+In the event of a chain fork, there may be multiple next blocks based on the
+current block, and multiple descendant chains. The chain contexts in forks are
+different, based on the different blocks in each fork. But they are based on
+the same pre-fork chain context.
+
+### Chain Context and Network Upgrades
+[chain-context-network-upgrades]: #chain-context-network-upgrades
+
+Each Network Upgrade includes some bilateral consensus rule changes. These
```

Comment:
what does bilateral mean here?

Reply:
It's a quote from the spec, I'll add a quick explanation and a reference to the spec:

```
Each network upgrade is introduced as a “bilateral consensus rule change”. In this kind of upgrade,
• there is an activation block height at which the consensus rule change takes effect;
• blocks and transactions that are valid according to the post-upgrade rules are not valid before the upgrade block height ;
• blocks and transactions that are valid according to the pre-upgrade rules are no longer valid at or after the activation block height .
```

https://zips.z.cash/protocol/protocol.pdf#networkupgrades

Reply:
Moved to #871.